drivers/display/display_ist3931.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright (c) 2024 Shen Xuyang <shenxuyang@shlinyuantech.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT ist_ist3931
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/drivers/display.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#define LOG_LEVEL CONFIG_DISPLAY_LOG_LEVEL
+LOG_MODULE_REGISTER(ist3931);
+
+#include "display_ist3931.h"
+
+typedef bool (*ist3931_bus_ready_fn)(const struct device *dev);
+typedef int (*ist3931_write_bus_fn)(const struct device *dev, uint8_t *buf, size_t len,
+				    bool command);
+
+union ist3931_bus {
+	struct i2c_dt_spec i2c;
+};
+
+struct ist3931_config {
+	union ist3931_bus bus;
+	ist3931_bus_ready_fn bus_ready;
+	struct gpio_dt_spec reset_gpio;
+	uint16_t height;
+	uint16_t width;
+	bool vc;      /* voltage_converter_circuits_enabled */
+	bool vf;      /* voltage_follower_circuits_enabled */
+	uint8_t bias; /* 0-7 */
+	uint8_t ct;   /* 0-255 */
+	uint8_t duty; /* 1-64 */
+	uint16_t fr;  /* frame_frequency_division */
+	bool shl;     /* 0 COM1->COMN 1 COMN->COM1 */
+	bool adc;     /* 0 seg1->seg132 1 seg132->seg1 */
+	bool eon;     /* 0 normal 1 Entire ON */
+	bool rev;     /* 0 RAM1->LCD ON 1 RAM0->LCD ON */
+	uint8_t x_offset;
+	uint8_t y_offset;
+};
+
+static int ist3931_write_bus_i2c(const struct device *dev, uint8_t *buf, bool command,
+				 uint32_t num_bytes)
+{
+	int ret;
+	const struct ist3931_config *config = dev->config;
+	const uint8_t command_byte = command ? IST3931_CMD_BYTE : IST3931_DATA_BYTE;
+	uint8_t *i2c_write_buf = k_malloc(num_bytes * 2);
```

Comment:
Thanks. I tried on a esp32s3 board but failed, so I checked `i2c_transfer `api and found this.
> In some cases drivers will generate separate transactions for each message fragment, with or without presence of [I2C_MSG_RESTART] in [flags].

This display controller need an i2c sequence like `START | ADDRESS BYTE|  COMMAND BYTE | ACK| DATA BYTE | ACK| COMMAND BYTE | ACK| DATA BYTE | STOP`. I think it means I can't use `i2c_transfer` to transfer the command byte and then use `i2c_transfer` to transfer the data byte, since the I2C_MSG_RESTART will interrupt the i2c sequence.

Reply:
Hmm, yeah- reading the API, this does seem to be the case. I see now that a control byte needs to be sent with *every* data byte, is that correct? I do think we need an intermediate buffer for this based on the way the I2C API is described. Perhaps we can statically allocate this buffer to be up to `IST3931_RAM_WIDTH/4` bytes, since it seems like that would be the largest transfer ever made to this display (from the `display_write` implementation, unless I have missed something)